lib/scopes/scopes_test.go

```diff
@@ -0,0 +1,650 @@
+// re-joining and re-segmenting produces the same segments.
+func TestDescendingSegments(t *testing.T) {
+	tts := []struct {
+		name     string
+		scope    string
+		segments []string
+	}{
+		{
+			name:     "root",
+			scope:    "/",
+			segments: nil,
+		},
+		{
+			name:     "empty",
+			scope:    "",
+			segments: nil,
+		},
+		{
+			name:     "single-segment",
+			scope:    "/aa",
+			segments: []string{"aa"},
+		},
+		{
+			name:     "multi-segment",
+			scope:    "/aa/bb/cc",
+			segments: []string{"aa", "bb", "cc"},
+		},
+		{
+			name:     "dangling separator single",
+			scope:    "/aa/",
+			segments: []string{"aa"},
+		},
+		{
+			name:     "dangling separator multi",
+			scope:    "/aa/bb/cc/",
+			segments: []string{"aa", "bb", "cc"},
+		},
+		{
+			name:     "leading empty segment",
+			scope:    "//aa/bb",
+			segments: []string{"", "aa", "bb"},
+		},
+		{
+			name:     "middle empty segment",
+			scope:    "/aa//bb",
+			segments: []string{"aa", "", "bb"},
+		},
+		{
+			name:     "trailing empty segment",
+			scope:    "/aa/bb//",
+			segments: []string{"aa", "bb", ""},
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			// verify that the iterator produces the expected segments
+			segs := seq.Collect(DescendingSegments(tt.scope))
+			require.Equal(t, tt.segments, segs)
+
+			// verfiy that joining and re-segmenting produces the same segments
+			segs2 := seq.Collect(DescendingSegments(Join(segs...)))
+			require.Equal(t, tt.segments, segs2)
+		})
+	}
+}
+
+// TestJoin tests the Join function for various combinations of segments, and verifies that re-segmenting
+// the joined value produces the original segments.
+func TestJoin(t *testing.T) {
+	tts := []struct {
+		name     string
+		segments []string
+		expect   string
+	}{
+		{
+			name:     "root",
+			segments: nil,
+			expect:   "/",
+		},
+		{
+			name:     "single segment",
+			segments: []string{"aa"},
+			expect:   "/aa",
+		},
+		{
+			name:     "multi segment",
+			segments: []string{"aa", "bb", "cc"},
+			expect:   "/aa/bb/cc",
+		},
+		{
+			name:     "leading empty segment preserved",
+			segments: []string{"", "aa", "bb"},
+			expect:   "//aa/bb",
+		},
+		{
+			name:     "middle empty segment preserved",
+			segments: []string{"aa", "", "bb"},
+			expect:   "/aa//bb",
+		},
+		{
+			name:     "trailing empty segment preserved",
+			segments: []string{"aa", "bb", ""},
+			expect:   "/aa/bb//",
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			// verify that Join produces the expected scope value
+			joined := Join(tt.segments...)
+			require.Equal(t, tt.expect, joined)
+
+			// verify that re-segmentation of the joined value produces the original segments
+			reSegmented := seq.Collect(DescendingSegments(joined))
+			require.Equal(t, tt.segments, reSegmented)
+		})
+	}
+}
+
+// TestCompare tests the Compare function for various combinations of scopes.
+func TestCompare(t *testing.T) {
```

Comment:
As with my comment re: GoDoc - it'd be useful to include a case for Ancestor/Descendent where they are not direct ancestors or direct descendents.

e.g:

```Go
		{
			name: "deeper ancestor",
			lhs:  "/aa/bb/cc",
			rhs:  "/aa",
			rel:  Ancestor,
		},
```